# Efetch Human and Chip-Seq data


In [2]:
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', 20)

In [3]:
#loading the ftp information from efetch table - Chip-Seq and Homo sapiens 
#without pathway
df = pd.read_csv('list_ftp_address.txt', sep=r'\s{2,}', engine='python', header=None, names=['ID'])

df['Address'] = df.ID.str.split("ftp:",expand=True)[1] #impossible split by GEO (There is no patters to GEO string )

#building the first part of address
df['Address'] = 'ftp:' + df['Address'].astype(str) + 'miniml/'
df.head()


ID  \
0  FTP download: GEO ftp://ftp.ncbi.nlm.nih.gov/g...   
1  FTP download: GEO ftp://ftp.ncbi.nlm.nih.gov/g...   
2  FTP download: GEO (WIG) ftp://ftp.ncbi.nlm.nih...   
3  FTP download: GEO (BED, BEDPE, HIC) ftp://ftp....   
4  FTP download: GEO (BED, BEDPE, HIC) ftp://ftp....   

                                             Address  
0  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE140nn...  
1  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...  
2  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nn...  
3  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...  
4  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...

In [6]:
#loading the Accession information from efetch table - GSEs

df_gse = pd.read_csv('list_access_efetch.txt', sep=r'\s{2,}', engine='python', header=None, names=['GSE_ID'])


df_gse['GSE'] = df_gse.GSE_ID.str.split('Accession: ', expand=True)[1]
df_gse['GSE'] = df_gse.GSE.str.split("\t", expand=True)[0]

df_gse_index = df_gse.reset_index(drop=True)
df_gse_index.head()

GSE_ID        GSE
0  Accession: GSE140664\tID: 200140664  GSE140664
1  Accession: GSE138674\tID: 200138674  GSE138674
2  Accession: GSE141497\tID: 200141497  GSE141497
3  Accession: GSE137662\tID: 200137662  GSE137662
4  Accession: GSE137661\tID: 200137661  GSE137661

In [7]:
df_gse_ftp = pd.concat([df_gse_index, df], axis=1)
df_gse_ftp['GSE'].nunique()
# df_gse_ftp.head()

5855

In [8]:
df_gse_ftp_final = df_gse_ftp[['GSE', 'Address']]
df_gse_ftp_final.head()

GSE                                            Address
0  GSE140664  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE140nn...
1  GSE138674  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...
2  GSE141497  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nn...
3  GSE137662  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...
4  GSE137661  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...

In [9]:

df_gse_ftp_final['Full_Address'] = df_gse_ftp_final['Address'].astype(str) + df_gse_ftp_final['GSE'] + '_family.xml.tgz'

df_gse_ftp_final.head()

/Users/gfrosi/opt/anaconda3/envs/lab-jacques/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


GSE                                            Address  \
0  GSE140664  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE140nn...   
1  GSE138674  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...   
2  GSE141497  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nn...   
3  GSE137662  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...   
4  GSE137661  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...   

                                        Full_Address  
0  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE140nn...  
1  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...  
2  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nn...  
3  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...  
4  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...

In [10]:
#saving this dataframe with information about GSE and their respective ftp address
df_gse_ftp_final.to_csv(r'/home/local/USHERBROOKE/frog2901/Documents/XML_TEST/GSE_FTP_Hs_ChipSeq.csv' , index = None, header=True)




FileNotFoundError: [Errno 2] No such file or directory: '/home/local/USHERBROOKE/frog2901/Documents/XML_TEST/GSE_FTP_Hs_ChipSeq.csv'

# Subprocess - Optimizing 


In [11]:
#subfile from final dataframe

df_gse_ftp_final_test_20rows = df_gse_ftp_final[:20]
df_gse_ftp_final_test_20rows


GSE                                            Address  \
0   GSE140664  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE140nn...   
1   GSE138674  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...   
2   GSE141497  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nn...   
3   GSE137662  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...   
4   GSE137661  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...   
5   GSE133072  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE133nn...   
6   GSE139810  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE139nn...   
7   GSE139685  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE139nn...   
8   GSE130810  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE130nn...   
9   GSE114981  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE114nn...   
10  GSE138812  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...   
11  GSE133208  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE133nn...   
12  GSE118993  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE118nn...   
13  GSE138378  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...   
14  GSE137527  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...   
15  GSE124197  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE124nn...   
16  GSE113654  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE113nn...   
17  GSE110405  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE110nn...   
18  GSE107315  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE107nn...   
19  GSE100789  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nn...   

                                         Full_Address  
0   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE140nn...  
1   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...  
2   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nn...  
3   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...  
4   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...  
5   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE133nn...  
6   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE139nn...  
7   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE139nn...  
8   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE130nn...  
9   ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE114nn...  
10  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...  
11  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE133nn...  
12  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE118nn...  
13  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE138nn...  
14  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE137nn...  
15  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE124nn...  
16  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE113nn...  
17  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE110nn...  
18  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE107nn...  
19  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nn...

In [20]:
#Get directories

import subprocess
import os

#get the files using the addres column and keep the code to adjust it here


#Funciton create directories to store our files

def create_dir(path):    
    try:
        os.mkdir(path)
        print("Directory {} done".format(path))
    except OSError:
        print("Directory {} failed".format(path))
            

root_path = '/Users/gfrosi/Documents/Frosi_lab_jacques/XML_TEST'
prefix_dir = 'GEO_'
counter = 1                                                                  

path = os.path.join(root_path , prefix_dir + str(counter))
create_dir(path)


#create the correct link and download the xml files

for index, row in df_gse_ftp_final_test_20rows.iterrows():    
    if (index + 1) % 10 == 0:  
        counter += 1
        path = os.path.join(root_path, prefix_dir + str(counter))
        create_dir(path)    
        
    sample_name = row["Address"].split("/")
    link = row["Address"] + sample_name[6] + "_family.xml.tgz" #we can keep in this form, or get the full address
    subprocess.run(["wget", link, "-P", path])
    

Directory /Users/gfrosi/Documents/Frosi_lab_jacques/XML_TEST/GEO_1 failed
Directory /Users/gfrosi/Documents/Frosi_lab_jacques/XML_TEST/GEO_2 failed
Directory /Users/gfrosi/Documents/Frosi_lab_jacques/XML_TEST/GEO_3 failed


In [13]:
#get the files using the Full_Addres column (the code to adjust the address was run before)


def create_dir(path):    
    try:
        os.mkdir(path)
        print("Directory {} done".format(path))
    except OSError:
        print("Directory {} failed".format(path))
            

root_path = '/home/local/USHERBROOKE/frog2901/Documents/XML_TEST'
prefix_dir = 'GEO_'
counter = 1                                                                  

path = os.path.join(root_path , prefix_dir + str(counter))
create_dir(path)


#create the correct link and download the xml files
for index, row in df_gse_ftp_final_test_20rows.iterrows():    
    if (index + 1) % 10 == 0:  
        counter += 1
        path = os.path.join(root_path, prefix_dir + str(counter))
        create_dir(path)    
        
    #sample_name = row["Address"].split("/")
    link = row["Full_Address"] #+ sample_name[6] + "_family.xml.tgz" #we can keep in this form, or get the full address
    subprocess.run(["wget", link, "-P", path])

Directory /home/local/USHERBROOKE/frog2901/Documents/XML_TEST/GEO_1 failed


FileNotFoundError: [Errno 2] No such file or directory: 'wget': 'wget'

# Get all XML files from general table  - 5855 GSE 